### PR TITLE
New App - Passbolt

### PR DESCRIPTION
**Pull request by josephcardillo**
This is the initial PR into develop for the Passbolt OCA.

### CODE REVIEW

File: apps/linode-marketplace-passbolt/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Make sure that the following values are updated at the top of the code block before running the commands:
+- TOKEN
+- ROOT_PASS
+
+SHELL:
+```
+export TOKEN="YOUR API TOKEN"
+export ROOT_PASS="aComplexP@ssword"
+export SOA_EMAIL_ADDRESS="email@domain.com"
+export POSTGRES_PASSWORD="aComplexP@ssword"
```
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/README.md:40]: remove postgres_password

File: apps/linode-marketplace-passbolt/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    }' \
+https://api.linode.com/v4/linode/instances
+```
+
+CLI:
+```
+export TOKEN="YOUR API TOKEN"
+export ROOT_PASS="aComplexP@ssword"
+export SOA_EMAIL_ADDRESS="email@domain.com"
+export POSTGRES_PASSWORD="aComplexP@ssword"
```
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/README.md:72]: remove postgres_password.

File: apps/linode-marketplace-passbolt/provision.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      lineinfile:
+        insertafter: EOF
+        path: group_vars/linode/vars
+        line: | 
+          passbolt_mysql_user: passboltadmin
+
+    - name: generating mysql root password
+      import_tasks: ../linode_helpers/roles/securemysql/tasks/passgen.yml
+
+    # - name: writing mysql root password to vars
```
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/provision.yml:46]: remove commented mysql pass

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    url: "https://github.com/passbolt/passbolt-dep-scripts/releases/latest/download/passbolt-ce-SHA512SUM.txt"
+    dest: "/tmp/passbolt-ce-SHA512SUM.txt"
+
+- name: Verify SHA-512 checksum
+  command: sha512sum -c /tmp/passbolt-ce-SHA512SUM.txt
+  args:
+    chdir: /tmp
+  register: checksum_result
+
+- name: Display checksum result
```
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:27]: remove debug task

File: apps/linode-marketplace-passbolt/roles/passbolt/templates/conf.j2 [hunk 5]
```diff
@@ -0,0 +1 @@
+# rename and add any templates for passbolt
```
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/roles/passbolt/templates/conf.j2:1]: this file can be removed if not required

File: apps/linode-marketplace-passbolt/README.md
Comment (hmorris3293) [inline apps/linode-marketplace-passbolt/README.md:1]: Readme needs to be updated for passbolt

Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/README.md:1]: README has been updated.

File: apps/linode-marketplace-passbolt/roles/passbolt/templates/conf.j2 (hunk 5 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/templates/conf.j2]: Removed.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml (hunk 4 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml]: Removed.

File: apps/linode-marketplace-passbolt/provision.yml (hunk 3 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/provision.yml]: Removed.

File: apps/linode-marketplace-passbolt/README.md (hunk 2 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/README.md]: Removed.

File: apps/linode-marketplace-passbolt/README.md (hunk 1 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/README.md]: Removed.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    echo "passbolt-ce-server passbolt/mysql-passbolt-username string passboltadmin" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/mysql-passbolt-password password {{ passbolt_db_admin_password }}" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/mysql-passbolt-password-repeat password {{ passbolt_db_admin_password }}" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/mysql-passbolt-dbname string passboltdb" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/nginx-configuration boolean true" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/nginx-configuration-three-choices select auto" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/nginx-domain string {{ _domain }}" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/nginx-certificate-file string /etc/letsencrypt/live/{{ _domain }}/fullchain.pem" | sudo debconf-set-selections
+    echo "passbolt-ce-server passbolt/nginx-certificate-key-file string /etc/letsencrypt/live/{{ _domain }}/privkey.pem" | sudo debconf-set-selections
+  become: yes
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:55]: let's try using the debconf module to keep this the Ansible way :)

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    - mailutils
+    state: present
+
+- name: updating postfix inet_interfaces
+  ansible.builtin.lineinfile:
+    path: /etc/postfix/main.cf
+    regexp: '^inet_interfaces ='
+    line: 'inet_interfaces = {{ ansible_default_ipv4.address }}'
+
+
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:77]: remove extra new line

File: apps/linode-marketplace-passbolt/roles/post/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    mode: '0600'
+    content: |
+      Sudo Username: {{ username }}
+      Sudo Password: {{ password }}
+      Passbolt Admin Database User: passboltadmin
+      Passbolt Database: passboltdb
+      Passbolt Database Password: {{ passbolt_db_admin_password }}
+      Mysql Root Password: {{ mysql_root_password }}
+      passbolt Login: admin
+      passbolt Password: {{ passbolt_password }}
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/post/tasks/main.yml:20]: I think this might need to be update and moved to the sudo user

**Review by n0vabyte — CHANGES_REQUESTED** [review]

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  get_url:
+    url: "https://download.passbolt.com/ce/installer/passbolt-repo-setup.ce.sh"
+    dest: "/tmp/passbolt-repo-setup.ce.sh"
+
+- name: Make Passbolt repository setup script executable
+  become: yes
+  become_user: root
+  file:
+    path: /tmp/passbolt-repo-setup.ce.sh
+    mode: "+x"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml]: Correction made.

**Comment by josephcardillo** [comment]
The requested changes have been made. The app deploys right to the passphrase screen, and mail can be sent via localhost. Best practices still needed, though, for mail deliverability to be ensured.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    mode: '0440'  
+    state: file
+
+- name: Set ownership and permissions for serverkey_private.asc
+  ansible.builtin.file:
+    path: /etc/passbolt/gpg/serverkey_private.asc
+    owner: www-data
+    group: www-data
+    mode: '0440'  
+    state: file
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:219]: we can consolidate this into a single task by using `loop` :)

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    gpg_secret_key: "{{ gpg_key_id_output.stdout }}"
+
+- name: Display GPG secret key variable
+  debug:
+    msg: "The GPG secret key ID is: {{ gpg_secret_key }}"
+
+- name: Writing gpg_secret_key to group_vars/linode/vars
+  lineinfile:
+    path: group_vars/linode/vars
+    line: "gpg_secret_key: '{{ gpg_secret_key }}'"
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:151]: we don't need this

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+  register: admin_registration_output
+
+- name: Extract registration URL
+  set_fact:
+    passbolt_admin_registration_url: "{{ admin_registration_output.stdout | regex_search('https://' + _domain | regex_escape + '/setup/start/[a-z0-9-]+/[a-z0-9-]+') }}"
+
+- name: Writing passbolt_admin_registration_url to group_vars/linode/vars
+  lineinfile:
+    path: group_vars/linode/vars
+    line: "passbolt_admin_registration_url: '{{ passbolt_admin_registration_url }}'"
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:203]: we don't need this

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml (hunk 11 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:151]: Removed.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml (hunk 12 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:203]: Removed.

File: apps/linode-marketplace-passbolt/provision.yml [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    - name: generating gpg_key_passphrase
+      set_fact:
+        gpg_key_passphrase: "{{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}"
+
+    - name: writing gpg_key_passphrase to vars
+      lineinfile:
+        insertafter: EOF
+        path: group_vars/linode/vars
+        line: | 
+          gpg_key_passphrase: {{ gpg_key_passphrase }}
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/provision.yml:23]: Good point! Updated.

File: apps/linode-marketplace-passbolt/provision.yml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  vars_files:
+    - group_vars/linode/vars
+  tasks:
+  # generate creds
+    - name: generating sudo user and password 
+      import_tasks: ../linode_helpers/roles/sudouser/tasks/main.yml
+      
+    - name: generating gpg_key_passphrase
+      set_fact:
+        gpg_key_passphrase: "{{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/provision.yml:16]: Removed and updated "writing gpg_key_passphrase to vars" task instead.

File: apps/linode-marketplace-passbolt/provision.yml [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    - name: generating passbolt_db_admin_password
+      set_fact:
+        passbolt_db_admin_password: "{{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}"
+
+    - name: writing passbolt_db_admin_password to vars
+      lineinfile:
+        insertafter: EOF
+        path: group_vars/linode/vars
+        line: | 
+          passbolt_db_admin_password: {{ passbolt_db_admin_password }}
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/provision.yml:34]: Updated.

File: apps/linode-marketplace-passbolt/provision.yml [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    - name: writing gpg_key_passphrase to vars
+      lineinfile:
+        insertafter: EOF
+        path: group_vars/linode/vars
+        line: | 
+          gpg_key_passphrase: {{ gpg_key_passphrase }}
+
+    - name: generating passbolt_db_admin_password
+      set_fact:
+        passbolt_db_admin_password: "{{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/provision.yml:27]: Nixed.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+- name: Download Passbolt repository setup script
+  get_url:
+    url: "https://download.passbolt.com/ce/installer/passbolt-repo-setup.ce.sh"
+    dest: "/tmp/passbolt-repo-setup.ce.sh"
+
+- name: Make Passbolt repository setup script executable
+  file:
+    path: /tmp/passbolt-repo-setup.ce.sh
+    mode: "0755"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:12]: Removed & get_url task updated with mode instead.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+    url: "https://github.com/passbolt/passbolt-dep-scripts/releases/latest/download/passbolt-ce-SHA512SUM.txt"
+    dest: "/tmp/passbolt-ce-SHA512SUM.txt"
+
+- name: Get SHA-512 checksum of the setup script
+  ansible.builtin.stat:
+    path: /tmp/passbolt-repo-setup.ce.sh
+    checksum_algorithm: sha512
+    get_checksum: yes
+  register: script_checksum
+
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:25]: Updated to use `assert` instead.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+  ignore_errors: yes
+
+- name: Remove setup script if checksum verification fails
+  ansible.builtin.file:
+    path: /tmp/passbolt-repo-setup.ce.sh
+    state: absent
+  when: script_checksum.stat.checksum != passbolt_checksum_value
+
+- name: Run passbolt-repo-setup.ce.sh script
+  command: "/tmp/passbolt-repo-setup.ce.sh"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:40]: True. Removed/updated.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,219 @@
+---
+# passbolt
+
+- name: Download Passbolt repository setup script
+  get_url:
+    url: "https://download.passbolt.com/ce/installer/passbolt-repo-setup.ce.sh"
+    dest: "/tmp/passbolt-repo-setup.ce.sh"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:7]: Done.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+- name: Get SHA-512 checksum of the setup script
+  ansible.builtin.stat:
+    path: /tmp/passbolt-repo-setup.ce.sh
+    checksum_algorithm: sha512
+    get_checksum: yes
+  register: script_checksum
+
+- name: Conditionally execute the script if checksum matches
+  ansible.builtin.shell:
+    cmd: sudo bash /tmp/passbolt-repo-setup.ce.sh
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:28]: True! Resolved.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+    executable: /bin/bash
+
+- name: Extract the key ID and save to a variable
+  shell: |
+    awk '/^sec/{getline; print $1}' /tmp/secret-keys.txt
+  register: gpg_key_id_output
+
+- name: Set the extracted key ID as a fact
+  set_fact:
+    gpg_secret_key: "{{ gpg_key_id_output.stdout }}"
```
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:142]: Fair enough!

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml (hunk 10 above)
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:219]: Done.

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml (hunk 22 above)
[the remark this answers is not in the document]
Comment (josephcardillo) [reply apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml]: Reworked with `slurp`

File: apps/linode-marketplace-passbolt/provision.yml [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+        path: group_vars/linode/vars
+        line: | 
+          passbolt_db_admin_password: {{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}
+
+    - name: writing passbolt_mysql_user to vars
+      lineinfile:
+        insertafter: EOF
+        path: group_vars/linode/vars
+        line: | 
+          passbolt_mysql_user: passboltadmin
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/provision.yml:33]:
```suggestion
    - name: writing deployment variables to vars file
      blockinfile:
        path: ./group_vars/galera/vars
        marker: "# {mark} INSTANCE VARS"
        block: |
          gpg_key_passphrase: {{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }} 
          passbolt_db_admin_password: {{ lookup('password', '/dev/null length=25 chars=ascii_letters,digits') }}          
          passbolt_mysql_user: passboltadmin
```

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+
+- name: Update GPG fingerprint in passbolt.php
+  ansible.builtin.replace:
+    path: /etc/passbolt/passbolt.php
+    regexp: "'fingerprint' => '.*',"
+    replace: "'fingerprint' => '{{ gpg_secret_key }}',"
+
+- name: Export GPG public key to serverkey.asc
+  shell: >
+    gpg --armor --export {{ gpg_secret_key }} > /etc/passbolt/gpg/serverkey.asc
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:154]:
```suggestion
  shell: "gpg --armor --export {{ gpg_secret_key }} > /etc/passbolt/gpg/serverkey.asc"
```

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+  args:
+    executable: /bin/bash
+
+- name: Export GPG private key to serverkey_private.asc
+  shell: >
+    gpg --armor --export-secret-keys {{ gpg_secret_key }} > /etc/passbolt/gpg/serverkey_private.asc
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:160]:
```suggestion
  shell: "gpg --armor --export-secret-keys {{ gpg_secret_key }} > /etc/passbolt/gpg/serverkey_private.asc"
```

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+  args:
+    executable: /bin/bash
+
+- name: Install Passbolt without admin user
+  become: yes
+  become_method: sudo
+  become_user: "{{ webuser }}"
+  shell: "/usr/share/php/passbolt/bin/cake passbolt install --no-admin --force"
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:168]:
```suggestion
  shell: /usr/share/php/passbolt/bin/cake passbolt install --no-admin --force
```

File: apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+- name: Verify SHA-512 checksum of the setup script
+  ansible.builtin.assert:
+    that:
+      - script_checksum.stat.checksum == passbolt_checksum_value
+    fail_msg: "Checksum verification failed for /tmp/passbolt-repo-setup.ce.sh"
+    success_msg: "Checksum verification passed for /tmp/passbolt-repo-setup.ce.sh"
+
+- name: Execute the passbolt setup script
+  ansible.builtin.shell:
+    cmd: bash /tmp/passbolt-repo-setup.ce.sh
```
Comment (n0vabyte) [inline apps/linode-marketplace-passbolt/roles/passbolt/tasks/main.yml:31]:
```suggestion
  ansible.builtin.shell: bash /tmp/passbolt-repo-setup.ce.sh
```

**Review by n0vabyte — CHANGES_REQUESTED** [review]
quick update to make things a bit shorter and cleaner.

**Comment by josephcardillo** [comment]
Updated one thing (var path in provision.yaml) after Elvis' requested changes. Re-deployed passbolt successfully.

**Review by hmorris3293 — APPROVED** [review]

**Review by n0vabyte — APPROVED** [review]

**Pull request merged**